Parallel up/start/stop

IlyaSkriblovsky opened the pull request:
Starting and stopping containers in dependency-ordered manner as proposed in #2619

- Starting is currently done sequentially in dependency order. This change makes it parallel while obeying dependencies. It may speed up starting by around 30% in my tests.
- Stopping is currently done all-parallel, regardless of dependencies. This change make stopping obey dependencies. This may slower shutting down, but makes it possible to stop processes gracefully .
- Killing behavior is not changed

Seems like `dep_ordered_op` is a superset of `parallel_op` (with `get_deps=lambda x: []`). Should I merge them together?

Ugly `silent` flag was used because `Service` does `log.info` on each started container and I don't want to interfere.

dnephin commented:
Impressive first draft, some high level comments:
* I think we should merge `dep_ordered_op` and `parallel_op` (as you suggest)
* `project.start()` is not actually what's called on `up`, there is a project.up()`. We might want to apply this to `up()` as well since that's what is most commonly used. I think start is rarely used.

IlyaSkriblovsky commented:
Thanks, forgot about `up`.

Am I correct, that test failures are not caused by my changes?

dnephin commented:
The `test_down` failure could be related if you changed the stdout of `project.stop()`, which I believe you've done.

IlyaSkriblovsky commented:
Ah, I've tested on Docker 1.9 on my machine and this test was skipped. Will fix that.

IlyaSkriblovsky commented:
I've tried to modify `up` to make it build, create and recreate containers in sequence, but to start them in parallel. But it turns out that I can't divide starting containers from creating them: for example if service `B` uses `container:A` network, creating `B` requires `A` to be already running.

So, I see two options:
- parallelize `up` by making it do all its work (build, create and start) in parallel (with dependency ordering, of course). But in that way the console output of building images will mix up.
- make it build when needs to be built in sequence and then create, recreate and start containers in parallel.

Do you agree that second option is the right way?

aanand commented:
Yes - I think all the building/pulling should happen first, and then (re)create+start can be as parallelised as possible.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "dependency-ordered-start-stop" git@github.com:IlyaSkriblovsky/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

IlyaSkriblovsky commented:
It seems like `test.acceptance.cli_test.py::CLITestCase::test_up_handles_sigint`/`sigterm` has a race condition. They start `up`, wait until there are expected number of containers started and then do `os.kill`. But `TopLevelCommand.up()` sets signal handlers only *after* it starts all containers (in `attach_to_logs()`). So, test method might send signal after containers is started, but before handlers is attached, and Compose process will be killed without proper shutdown, leaving containers running.

This issue is outside of this PR and not caused by my changes. But seems like my changes affected some timings and this issue now appears with greater probability than before (it appeared only in py27 in last jenkins build, but not in py34, for example).

I added a workaround just to make tests to pass by adding `, delay=1)` to `wait_on_condition` calls to make them send signals not earlier than after 1 second. This might be removed from final code.

dnephin commented:
I'll work on a fix for that issue

IlyaSkriblovsky commented:
What about the rest of PR? :)

dnephin commented:
Oops, I guess I had a "fixes (comment)" in another PR, and I guess that caused github to close this issue.

The test race condition should be fixed on master now

IlyaSkriblovsky commented:
Removed my workaround for #2784 and rebased on master

IlyaSkriblovsky commented:
rebased on master

dnephin commented on tests/acceptance/cli_test.py:
```diff
@@ -1227,7 +1228,7 @@ def test_events_json(self):
         result = wait_on_process(events_proc, returncode=1)
         lines = [json.loads(line) for line in result.stdout.rstrip().split('\n')]
-        assert [e['action'] for e in lines] == ['create', 'start', 'create', 'start']
+        self.assertEqual(Counter(e['action'] for e in lines), {'create': 2, 'start': 2})
```
We prefer `assert` over `self.assertEqual`. `pytest` takes care of formatting the error properly, and generally does a better job than unittest at formatting it, and displaying relevant information.

dnephin commented on compose/parallel.py:
```diff
@@ -133,3 +92,86 @@ def parallel_kill(containers, options):
+    """
+    objects = list(objects)
+    stream = get_output_stream(sys.stderr)
+    writer_class = DummyParallelStreamWriter if silent else ParallelStreamWriter
```
I don't think we need to worry about being silent. We can have start and up print a similar message. If they already have logging somewhere else we can move it here.

dnephin commented:
I think  the parallel function has changed a bit since this PR was first prepared. Some logic was extracted into a separate function to reduce the overall complexity, and I think the error handling was simplified a bit.

I think it would be good if we could keep as much as the existing implementation as possible, and just add a few args for handling dependencies. The big difference with this implementation is the `feed()` function, right? Otherwise it's approximately the same?

IlyaSkriblovsky commented:
Thanks for feedback. I will rebase this and try to extract queue processing from `dep_ordered_op` just like it was done to `parallel_op`. Lets see if it will be possible to share other parts of the function.

IlyaSkriblovsky commented:
Squashed and rebased onto master.

1. I've refactored my `dep_ordered_op` to be compatible with `parallel_execute` and to inherit its structure and some code parts.

2. New `parallel_execute` still handles silent mode (enabled by `msg=None`). This is to make it silent for `up` because output of `up` differs from output of other commands. Each line of `up` output may have different operation name (Starting/Creating/Recreating/Up-to-date). This is actually not a problem: we might replace `msg` argument of `parallel_execute` with function returning operation name for each object (service). But anyways, it will require removing logging from multiple methods of `Service` class: `execute_convergence_plan`, `create_container`, `recreate_container`, etc, or passing some sort of silence flag via arguments. Does it worth it? Please advice.

3. #2804 that was merged today adds test case which tests that `up` raises `APIError` on some broken config. This testcase is currently failing on this branch, because `up` was redone by using `parallel_execute` which consumes `APIError` and just logs them to console. What is the correct expected behavior of `up`? To reraise `APIError` like this testcase expects or to consume `APIError` just like other methods (`stop`/`restart`/`pause`) that was implemented by `parallel_execute`?

dnephin commented:
For 2 I agree about the logging with `up`. The way you have it now seems correct.

For 3, I think the test case will have to change. We can probably test that behaviour in a better way.

IlyaSkriblovsky commented:
Fixed test by checking `len(up()) == 0` (`up()` returns a list of successfully started containers)

IlyaSkriblovsky commented:
Weird, `tests/integration/project_test.py::ProjectTest::test_up_with_network_static_addresses` seems to pass well on my machine...

IlyaSkriblovsky commented:
I made temporary commit that makes `up` not to swallow APIErrors. Hope this will help to understand why `tests/integration/project_test.py::ProjectTest::test_up_with_network_static_addresses` is failing. Will reset this commit after that.

IlyaSkriblovsky commented:
Ok, I played a bit with failing `test_up_with_network_static_addresses` test, then reverted temporary commits back and now it suddenly passes :-/  So, it seems this test is somewhat unstable and this is not related to this PR.

So, can I do anything else for this PR?

aanand commented:
I've checked it out and it works nicely. I think the only issue I have is with output - we need to do some refactoring so that the output of `up` ("Recreating X", "Y is up-to-date" etc) is handled in the same way as that of `start`, `stop`, `kill` etc. Otherwise we could easily get jumbled output. But we can do that separately.

LGTM

dnephin commented on compose/project.py:
```diff
@@ -200,13 +201,40 @@ def get_network_mode(self, service_dict, networks):
         return containers
 
     def stop(self, service_names=None, **options):
-        parallel.parallel_stop(self.containers(service_names), options)
```
I think `parallel_stop()` can be deleted now. I don't think it's used anywhere else

dnephin commented:
LGTM, just one minor request to remove the old unused function

IlyaSkriblovsky commented:
Removed unused `parallel_stop` and rebased to master
